File: pe/nonvisualstudio/elixir/Implementation/Problem0008.ex
defmodule Problem0008 do
  def importProblemSet(path) do
    File.read!(path)
    |> String.replace("\n", "")
    |> String.replace("\r", "")
  end

  def multiplyDigitsOfString(string) do
    String.codepoints(string)
    |> Enum.map(fn(elem) -> Integer.parse(elem) end)
    |> Enum.map(fn(item) -> elem(item,0) end)
    |> Enum.reduce(fn(elem, acc) -> elem * acc end)
  end

  def multiplyPortionOfString(howLong, string), do: String.slice(string, 0, howLong) |>  multiplyDigitsOfString

  def maxOfMultiplesOfString("", _, result), do: result
  def maxOfMultiplesOfString(string, howLong, result) do
    cond do
      result > multiplyPortionOfString(howLong, string) ->
        maxOfMultiplesOfString(restOfString(string), howLong, result)
      true ->
        maxOfMultiplesOfString(restOfString(string), howLong, multiplyPortionOfString(howLong, string))
    end
  end

  def maxOfMultiplesOfTxtFile(howLong, path) do
    importProblemSet(path)
    |> maxOfMultiplesOfString(howLong, 0)
  end

  defp restOfString(string), do: String.slice(string, 1..(String.length(string) - 1))
end

File: pe/nonvisualstudio/clojure/Implementation/Problem0008.clj
(ns pe.clojure.implementation.problem0008)

(defn importProblemSet [path]
  (apply str
    (->>
      (slurp path)
      (filter (fn [x] (Character/isDigit x))))))

(defn multiplyDigitsOfString [string]
  (reduce *
    (->>
      (seq string)
      (map int)
      (map (fn [elem] (- elem 48))))))

(defn multiplyDigitsOfPotion [howLong string]
  (->>
    (take howLong string)
    (multiplyDigitsOfString)))

(defn maxOfMultiplesOfString [howLong string]
  (loop [currentString string
         result 0]
    (cond
      (= currentString "") 
        result
      (> (multiplyDigitsOfPotion howLong currentString) result)
        (recur (subs currentString 1) (multiplyDigitsOfPotion howLong currentString))
      :else 
        (recur (subs currentString 1) result))))

(defn maxOfMultiplesOfTxtFile [howLong path]
  (maxOfMultiplesOfString
    howLong
    (importProblemSet path)))
